lib/puppet_x/puppetlabs/scheduled_task/taskscheduler2.rb

```diff
@@ -0,0 +1,318 @@
+    task_path.rpartition('\\')[2]
+  end
+
+  # Returns an array of scheduled task names.
+  # By default EVERYTHING is enumerated
+  # option hash
+  #    include_child_folders: recurses into child folders for tasks. Default true
+  #    include_compatibility: Only include tasks which have any of the specified compatibility levels. Default empty array (everything is permitted)
+  #
+  def self.enum_task_names(folder_path = ROOT_FOLDER, options = {})
+    raise TypeError unless folder_path.is_a?(String)
+
+    options[:include_child_folders] = true if options[:include_child_folders].nil?
+    options[:include_compatibility] = [] if options[:include_compatibility].nil?
+
+    array = []
+
+    task_folder = task_service.GetFolder(folder_path)
+    task_folder.GetTasks(TASK_ENUM_HIDDEN).each do |task|
+      included = true
+      included = included && options[:include_compatibility].include?(task.Definition.Settings.Compatibility) unless options[:include_compatibility].empty?
+
+      array << task.Path if included
+    end
+    return array unless options[:include_child_folders]
+
+    task_folder.GetFolders(0).each do |child_folder|
+      array += enum_task_names(child_folder.Path, options)
+    end
+
+    array
+  end
+
+  def self.task(task_path)
+    raise TypeError unless task_path.is_a?(String)
+
+    task_folder = task_service.GetFolder(folder_path_from_task_path(task_path))
+
+    task_object = task_folder.GetTask(task_name_from_task_path(task_path))
+
+    task_object
+  end
+
+  def self.definition(task_object = nil)
+    definition = task_service.NewTask(0)
+    definition.XmlText = task_object.XML unless task_object.nil?
+
+    definition
+  end
+
+  # Creates or Updates an existing task with the supplied task definition
+  # If task_object is a string then this is a new task and the supplied object is the new task's full path
+  # Otherwise we expect a Win32OLE Task object to be passed through
+  def self.save(task_object, definition, password = nil)
+    task_path = task_object.is_a?(String) ? task_object : task_object.Path
+
+    task_folder = task_service.GetFolder(folder_path_from_task_path(task_path))
+    task_user = nil
+    task_password = nil
+
+    case definition.Principal.LogonType
+      when TASK_LOGON_PASSWORD, TASK_LOGON_INTERACTIVE_TOKEN_OR_PASSWORD
+        task_user = definition.Principal.UserId
+        task_password = password
+    end
+    task_folder.RegisterTaskDefinition(task_name_from_task_path(task_path),
+                                       definition, TASK_CREATE_OR_UPDATE, task_user, task_password,
+                                       definition.Principal.LogonType)
+  end
+
+  # Delete the specified task name.
+  #
+  def self.delete(task_path)
+    raise TypeError unless task_path.is_a?(String)
+    task_folder = task_service.GetFolder(folder_path_from_task_path(task_path))
+
+    result = task_folder.DeleteTask(task_name_from_task_path(task_path),0)
```

Comment:
COM wrappers automatically raised for HRESULT < 0 failures on method calls like this (like https://github.com/puppetlabs/puppet/blob/master/lib/puppet/util/windows/taskscheduler.rb#L265) with a `Puppet::Util::Windows::Error`.
https://github.com/puppetlabs/puppet/blob/master/lib/puppet/util/windows/com.rb#L141-L143

I *think* we should probably continue to do the same here?

Reply:
Yeah...as these are WIN32OLERuntimeError not COM errors things are not as clear cut.  The WIN32OLERuntimeError stuff is horrible to work with.

Reply:
Yeah, I now recall our discussion about the `WIN32OLERuntimeError` business and the lack of error codes as an attribute of the error instance... so that's a problem for sure.

But in particular I was looking at the ` == Puppet::Util::Windows::COM::S_OK`, which I assumed indicated you got back a `HRESULT`... However, looking at the docs now, I don't think that that's right - `DeleteTask` says `this method does not return a value`, so I think if nothing is raised you just `return true` at the end:

https://msdn.microsoft.com/en-us/library/windows/desktop/aa382566%28v=vs.85%29.aspx?f=255&MSPPError=-2147217396

This is confirmed by the COM interface docs as well https://msdn.microsoft.com/en-us/library/windows/desktop/aa381345(v=vs.85).aspx

Note: All COM interface methods return a `HRESULT` status code by convention, and if they have any return values from the method they're actually `[out]` parameters to pointer-pointers in the IDL.  I assume `WIN32OLERuntimeError` probably already does the `FAILED` check under the hood for us, so my original point is moot I think.

Reply:
Additional note: I am now a little concerned that this method is always returning `false`, yet we have no test failures to show for it.

Reply:
Weird.  competing docs -
https://msdn.microsoft.com/en-us/library/windows/desktop/aa381345(v=vs.85).aspx

```
Return value
If this method succeeds, it returns S_OK. Otherwise, it returns an HRESULT error code.
```

Reply:
Modified delete method
```
  def self.delete(task_path)
    raise TypeError unless task_path.is_a?(String)
    task_folder = task_service.GetFolder(folder_path_from_task_path(task_path))

    task_folder.DeleteTask(task_name_from_task_path(task_path),0)
  end
```

and tests
```
    it 'should delete a task that exists' do
...
    it 'should raise an error for a task that does not exist' do
...
```

Reply:
Just for the sake of anyone else following the thread, the docs are not competing, just  targeted at different consumers.  COM methods all returns status HRESULTs as mentioned in https://github.com/puppetlabs/puppetlabs-scheduled_task/pull/6#discussion_r160872843

For COM functions to return, for instance, new objects, they have additional `[out]` parameters that are passed into the function as pointer pointers... For instance,

```cpp
HRESULT GetFolder(
  [in]  BSTR        path,
  [out] ITaskFolder **ppFolder
);
```

Essentially you pass the address of your `ITaskFolder *` (interface pointer), so that the calling function can write the actual pointer to the memory address you've given when it creates the output object.

Because throwing C++ exceptions out of COM functions is not supported (what would catching  a C++ exception in Visual Basic even mean?), status is conveyed by the convention of always returning a `HRESULT` value for success / fail.

When we use `WIN32OLE` in Ruby, we're using the scripting compatible interfaces, which support late-binding and going through the `IDispatch` interface... so when a function has no `[out]` params in the actual IDL, that means it's basically a `void`.  But the easiest thing to do is just look at the "scripting" docs at https://msdn.microsoft.com/en-us/library/windows/desktop/aa383607(v=vs.85).aspx rather than the C++ docs.